### ISSUE TITLE
Port-restricted IP addresses do not work with VLAN tagging

### ISSUE THREAD

**Issue by andywingo** [Igalia/snabbswitch#98]

**Comment by dpino** [comment]
So far what I have done is to clone the end-to-end tests so packets and configuration files have VLAN tagging activated #108 . There are some tests that pass and others don't. Code is merged in lwaftr_olive now (end-to-end-vlan.sh).

Many of the tests that don't pass are related with ICMP packets. I believe #99 will fix those. So I'm focusing instead on those which are not ICMP related. My next step is to go through every failing test and see what are the differences between packets generated.

With regard to spotting the bug directly in the source code, I didn't try that much.

**Comment by dpino** [comment]
One failing test is:

```
Testing: from-internet IPv4 packet found in the binding table, needs IPv6 fragmentation (3).
done
../data/vlan/tcp-afteraftr-ipv6-3frags.pcap /tmp/endoutv6.pcap differ: byte 33, line 1
```

**tcp-afteraftr-ipv6-3frags.pcap** headers
```
$ tcpdump -e -qns 0 -r ../data/vlan/tcp-afteraftr-ipv6-3frags.pcap
reading from file ../data/vlan/tcp-afteraftr-ipv6-3frags.pcap, link-type EN10MB (Ethernet)
01:00:00.000000 22:22:22:22:22:22 > 44:44:44:44:44:44, 802.1Q, length 1282: vlan 1638, p 0, ethertype IPv6, 8:9:a:b:c:d:e:f > 127:11:12:13:14:15:16:128: frag (0|1216) truncated-ip - 1504 bytes missing! 1.2.3.4.2300 > 178.79.150.233.2700: tcp 2680
01:00:00.000000 22:22:22:22:22:22 > 44:44:44:44:44:44, 802.1Q, length 1282: vlan 1638, p 0, ethertype IPv6, 8:9:a:b:c:d:e:f > 127:11:12:13:14:15:16:128: frag (1216|1216)
01:00:00.000000 22:22:22:22:22:22 > 44:44:44:44:44:44, 802.1Q, length 354: vlan 1638, p 0, ethertype IPv6, 8:9:a:b:c:d:e:f > 127:11:12:13:14:15:16:128: frag (2432|288)
```

**/tmp/endoutv6.pcap** headers
```
$ tcpdump -e -qns 0 -r /tmp/endoutv6.pcap
reading from file /tmp/endoutv6.pcap, link-type EN10MB (Ethernet)
01:00:00.000000 22:22:22:22:22:22 > 44:44:44:44:44:44, 802.1Q, length 1274: vlan 1638, p 0, ethertype IPv6, 8:9:a:b:c:d:e:f > 127:11:12:13:14:15:16:128: frag (0|1208) truncated-ip - 1512 bytes missing! 1.2.3.4.2300 > 178.79.150.233.2700: tcp 2680
01:00:00.000000 22:22:22:22:22:22 > 44:44:44:44:44:44, 802.1Q, length 1274: vlan 1638, p 0, ethertype IPv6, 8:9:a:b:c:d:e:f > 127:11:12:13:14:15:16:128: frag (1208|1208)
01:00:00.000000 22:22:22:22:22:22 > 44:44:44:44:44:44, 802.1Q, length 370: vlan 1638, p 0, ethertype IPv6, 8:9:a:b:c:d:e:f > 127:11:12:13:14:15:16:128: frag (2416|304)
```
The difference is the truncation of the first packet. It's one byte shorter (1208 bytes vs 1216 bytes). The diff reports a difference in byte 33, but I think that's PCAP header control information because it's data before the payload.

**tcp-afteraftr-ipv6-3frags.pcap**
```
0000000: d4c3 b2a1 0200 0400 0000 0000 0000 0000  ................
0000010: ffff 0000 0100 0000 0000 0000 0000 0000  ................
0000020: fa04 0000 fa04 0000 4444 4444 4444 2222  ........DDDDDD""
0000030: 2222 2222
```

**/tmp/endoutv6.pcap**
```
0000000: d4c3 b2a1 0200 0400 0000 0000 0000 0000  ................
0000010: ffff 0000 0100 0000 0000 0000 0000 0000  ................
0000020: 0205 0000 0205 0000 4444 4444 4444 2222  ........DDDDDD""
0000030: 2222 2222
```

**Comment by kbara** [comment]
I'm explicitly not addressing fragmentation problems until fixing  #100 . If this persists after #100 is fixed, ping me. Byte 33 is indeed part of the pcap header.

**Comment by andywingo** [comment]
After all the bug fixes we are not able to reproduce this issue.  We're asking for more information but in the meantime we'll roll a v1.1 with all we have.  Reassigning to v1.2.

**Comment by kbara** [comment]
For future reference: the bug @dpino mentioned is fixed, via https://github.com/Igalia/snabbswitch/issues/100 (redefining MTUs to be L3), which depends on https://github.com/Igalia/snabbswitch/pull/105 (vlan/header handling refactorings). I don't think it's related to the externally-reported bug.

**Comment by kbara** [comment]
Can we close this? I believe the vlan tests (and associated code fixes) we did for 1.1 (and on lwaftr_olive) resolve every issue related to this that we can reproduce, although not every issue that has been reported. Thoughts, @andywingo , @dpino ?

**Comment by dpino** [comment]
@kbara SGTM. It seems most of the cases, if not all, are solved now. If there are new bugs related with this issue in the future, we can open a new issue with a narrower scope.
